File: apps/api/src/services/get-quizzes.ts
import { db } from "@/lib/db"
import { Request, Response } from "express"
import { sql } from "kysely"
import { getUserId } from "./get-user-id"

export const getQuizzes = async (req: Request, res: Response) => {
  const userId = await getUserId(req)

  const quizzes = await db
    .selectFrom('quizzes')
    .leftJoin('user_quiz_attempts', 'quizzes.id', 'user_quiz_attempts.quiz_id')
    .select([
      'quizzes.id',
      'quizzes.question',
      'quizzes.difficulty',
      'quizzes.explanation',
      'quizzes.options',
      'quizzes.created_at',
      'quizzes.topic',
      db.fn.count('user_quiz_attempts.id').as('total_attempt_count'),
      sql<string>`MAX(CASE WHEN user_quiz_attempts.user_id = ${userId || null} THEN user_quiz_attempts.attempt_date ELSE NULL END)`.as('user_last_attempt_date')
    ])
    .groupBy('quizzes.id')
    .execute()

  const quizzesWithAttemptInfo = quizzes.map(quiz => {
    const total_attempt_count = Number(quiz.total_attempt_count)
    const user_last_attempt_date = quiz.user_last_attempt_date

    return {
      ...quiz,
      total_attempt_count,
      ...(userId ? {
        has_been_practiced: !!user_last_attempt_date,
        last_attempt_date: user_last_attempt_date
      } : {})
    }
  })

  return res.json(quizzesWithAttemptInfo)
}

File: apps/backend/backend/services/get_quizzes.py
from flask import jsonify, request
from sqlalchemy import func
from sqlalchemy.orm import aliased
from backend.models.quiz import Quiz
from backend.models.user_quiz_attempt import UserQuizAttempt
from backend.services.get_user_id import get_user_id
from backend.libs.db import get_db
from sqlalchemy import case

def get_quizzes():
    db = get_db()
    user_id = get_user_id(request)
    
    uqa = aliased(UserQuizAttempt)

    quizzes = db.query(
        Quiz,
        func.count(uqa.id).label('total_attempt_count'),
        func.max(case(
            (uqa.user_id == user_id, uqa.created_at),
            else_=None
        )).label('user_last_attempt_date')
    ).outerjoin(uqa, Quiz.id == uqa.quiz_id).group_by(Quiz.id).all()
    
    quizzes_with_attempt_info = []
    for quiz, total_attempt_count, user_last_attempt_date in quizzes:
        quiz_dict = {
            'id': quiz.id,
            'question': quiz.question,
            'difficulty': quiz.difficulty,
            'explanation': quiz.explanation,
            'options': quiz.options,
            'created_at': quiz.created_at,
            'topic': quiz.topic,
            'total_attempt_count': total_attempt_count
        }
        
        if user_id:
            quiz_dict.update({
                'has_been_practiced': bool(user_last_attempt_date),
                'last_attempt_date': user_last_attempt_date
            })
        
        quizzes_with_attempt_info.append(quiz_dict)

    return jsonify(quizzes_with_attempt_info)
